Build tool & testing framework.

joshlasdin opened the pull request:
Put together some basic stuff for building & running tests:

* Added webpack (http://webpack.github.io/) for transpiling everything to a single file (currently outputs to `dist/bundle.js`)
* Build script can be run with `npm run build`, automatically builds on install (that way we can keep dist files out of the repo)
* Also set up a karma test runner to run a mocha suite in PhantomJS. This allows us to run our unit tests in a real (albeit headless) browser, which is potentially where this will be used in the endrun. Also allows us to run the tests in other browsers later if we want. Tests can be run with `npm test`.
* Went with mocha/chai/sinon for mocha's excellent promise support. After doing a little research on handling promises with jasmine, it seemed a little cumbersome to work with. The assertion APIs look pretty similar so hopefully it won't be too frustrating of a switch. With using karma, we can always change it out later.

Once we start getting some tests together for stuff it might be nice to have things running through something like TravisCI. I've set up karma to output some lcov data, so we can integrate with something like coveralls at a later point too. No sense really setting it up now as it has nothing to do haha.

Give it a look-over and let me know what you think, was going to post some discussions on it first, but seemed easy enough to just put together and go from there.

joshlasdin commented:
Related to: https://trello.com/c/PqECCaOX/7-build-tools

epixa commented on package.json:
```diff
@@ -15,5 +21,25 @@
+  "scripts": {
+    "install": "npm run build",
+    "build": "./node_modules/webpack/bin/webpack.js",
+    "test": "./node_modules/karma/bin/karma start ./karma.config.js"
```
`npm` is smart enough to handle module bin paths for you, so you can just do:

```
  "build": "webpack",
  "test": "karma start ./karma.config.js"
```

joshlasdin replied:
Interesting, thought I tried that and had an issue, but will double check. That would definitely be a lot cleaner.

epixa commented:
This is awesome.

One thing I want to make sure we cover sooner rather than later is making sure the unit tests can run both through node.js _and_ through the browser since JSClient isn't a browser-only thing.  That's in no way a blocker for this PR though.

joshlasdin commented:
@epixa Great point, I'll see what I can do about adding a node runner in as well. Should be easy enough.

joshlasdin commented:
@epixa Alright, think I got this to a better spot. Here's a summary of the new revisions:
- Added a transpile script, which just runs everything in `lib/` through babel and dumps it in `build/` for consumption in node projects
- Tests now run in both node & browser
- Added a minified version of the bundle for use in browsers
- Set up a prepublish script & `.npmignore` so whenever this gets pushed to the npm registry it compiles the consumable files and ignores the source es6
- Updated the README to describe some of the stuff added in this PR

Let me know what you think!

epixa commented:
Great stuff!
